### PR TITLE
feat: add Huly cloud (huly.app) support

### PR DESCRIPTION
**Pull request by fabioklr**
## Summary

`huly-cli` currently only works with self-hosted Huly instances. This PR adds full support for **Huly cloud (huly.app)** and an OTP login command:

- **Accounts endpoint**: Cloud uses `account.huly.app` subdomain instead of `/_accounts` reverse-proxy path. Auto-detected from the configured URL.
- **RPC param format**: Cloud expects named params (`{"email": ...}`) instead of positional arrays. `selectWorkspace` uses `workspaceUrl` key instead of a positional slug.
- **Transactor URL**: Cloud returns a per-region transactor in the `selectWorkspace` response (e.g. `wss://europe-tr1.huly.app`). The `/_transactor` path returns the SPA on cloud. The transactor base is now derived from the response and persisted in the auth cache.
- **OTP login (`auth login-otp`)**: Cloud accounts using magic-link login have password login locked (`PasswordLoginLocked`). The new command sends an OTP code to the user's email and validates it — also resets any failed-login lockout.

All changes are **backward-compatible** with self-hosted instances.

## Files changed

| File | Change |
|------|--------|
| `auth.py` | Cloud endpoint detection, named params, dynamic transactor URL, `send_otp`/`login_otp`/`_complete_login` refactor |
| `client.py` | Use cached `transactor_base` from auth instead of hardcoded `/_transactor` |
| `auth_cmd.py` | New `login-otp` command with interactive code prompt |
| `config.py` | Add `otp_code` to `HulyConfig`, `transactor_base` to `AuthCache` (with persistence) |

## Test plan

- [x] `huly auth login-otp` sends OTP, prompts for code, authenticates successfully
- [x] `huly auth status` shows authenticated state
- [x] `huly projects list` returns projects
- [x] `huly issues list --project <ID>` returns issues  
- [x] `huly documents list` returns documents
- [x] `huly --json projects list` returns valid JSON
- [ ] Verify self-hosted instances still work (no access to one — changes are gated by URL detection)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/huly_cli/auth.py [hunk 1]
```diff
@@ -3,21 +3,60 @@
...
+        headers["Authorization"] = f"Bearer {token}"
+    resp = await http.post(
+        url,
+        headers=headers,
+        json={"id": request_id, "method": method, "params": params},
+    )
+    resp.raise_for_status()
+    body = resp.json()
+    if "error" in body:
+        raise AuthError(f"{method} failed: {body['error']}")
```
Comment (teslakoile) [inline src/huly_cli/auth.py:64]: If the server returns `{"error": null}` this raises `AuthError: "login failed: None"`. Guard with `if "error" in body and body["error"] is not None`.

File: src/huly_cli/auth.py [hunk 2]
```diff
@@ -27,93 +66,136 @@ async def login(config: HulyConfig) -> AuthCache:
...
-        )
-        resp.raise_for_status()
-        body = resp.json()
-        if "error" in body:
-            raise AuthError(f"Login failed: {body['error']}")
+        body = await _rpc(http, accounts_url, "login", {
+            "email": config.email,
+            "password": config.password,
+        })
         account_token: str = body["result"]["token"]
```
Comment (teslakoile) [inline src/huly_cli/auth.py]: Bare `body["result"]` — `KeyError` if the API returns a 200 without a `result` key (version mismatch, proxy, etc). Check presence first and raise `AuthError` with a descriptive message.

**Review by teslakoile — COMMENTED** [review]
Great feature — cloud support + OTP is valuable. A few things to address before this can merge:

**Merge conflict (blocker):** `auth.py` was significantly changed on master since this branch was cut. Needs a rebase before CI will run.

**No tests:** The new `_accounts_url`, `_rpc`, `_complete_login`, `send_otp`, `login_otp` paths and the `login-otp` command have no automated tests. See `tests/test_auth.py` and `tests/test_auth_cmd.py` for the existing patterns to follow.

Inline comments below on the specific code issues.

File: src/huly_cli/auth.py [hunk 3]
```diff
@@ -27,93 +66,136 @@ async def login(config: HulyConfig) -> AuthCache:
...
-            accounts_url,
-            json={"id": "1", "method": "login", "params": [config.email, config.password]},
-        )
-        resp.raise_for_status()
-        body = resp.json()
-        if "error" in body:
-            raise AuthError(f"Login failed: {body['error']}")
+        body = await _rpc(http, accounts_url, "login", {
+            "email": config.email,
+            "password": config.password,
```
Comment (teslakoile) [inline src/huly_cli/auth.py]: Named params are used unconditionally here (and in `selectWorkspace` / `getUserWorkspaces` below). Self-hosted instances expect positional arrays: `[email, password]`. The branch description notes self-hosted wasn't verified — this is likely why. Consider detecting param format via `_accounts_url` / URL detection the same way you detect the endpoint.

File: src/huly_cli/auth.py [hunk 4]
```diff
@@ -27,93 +66,136 @@ async def login(config: HulyConfig) -> AuthCache:
...
+    account_token: str,
+) -> AuthCache:
+    """Complete login after obtaining an account token (from password or OTP)."""
+    # Select workspace
+    body = await _rpc(
+        http, accounts_url, "selectWorkspace",
+        {"workspaceUrl": config.workspace, "kind": "external"},
+        token=account_token, request_id="2",
+    )
+    ws_result = body["result"]
```
Comment (teslakoile) [inline src/huly_cli/auth.py]: Same bare `body["result"]` issue — `KeyError` if absent. Should raise a clean `AuthError`.

File: src/huly_cli/auth.py [hunk 5]
```diff
@@ -27,93 +66,136 @@ async def login(config: HulyConfig) -> AuthCache:
...
+
+    # Cloud returns a per-region transactor endpoint (e.g. wss://europe-tr1.huly.app).
+    # Use it for REST calls instead of the default /_transactor path.
+    endpoint = ws_result.get("endpoint", "")
+    if endpoint:
+        transactor_base = endpoint.replace("wss://", "https://").replace("ws://", "http://")
+
+    # Get workspace UUID via getUserWorkspaces
+    body = await _rpc(
+        http, accounts_url, "getUserWorkspaces", {},
```
Comment (teslakoile) [inline src/huly_cli/auth.py]: Self-hosted `getUserWorkspaces` expects positional params (`[]`), not a named dict (`{}`). Same positional vs named params issue as above.

**Comment by fabioklr** [comment]
Thanks for the review, Kyle — all points addressed in the force-push (commit 3855cd8):

**Rebased** onto v0.1.4. Clean now.

**`_rpc` error guard** — now `if "error" in body and body["error"] is not None`, so `{"error": null}` no longer raises.

**Bare `body["result"]`** — replaced with `_require_result_dict` / `_require_str` helpers that raise descriptive `AuthError`s in login, `selectWorkspace`, and `_complete_login`.

**Positional vs named params** — added `_is_cloud(config)` + `_params(cloud=..., self_hosted=...)` helpers. Cloud sends named dicts; self-hosted sends positional arrays. Applied to `login`, `selectWorkspace`, `getUserWorkspaces`, `validateOtp`, and `loginOtp`.

**Tests** — added `tests/test_auth_cloud.py` (14 tests): endpoint selection, per-RPC param-format detection for both cloud and self-hosted, `_rpc` null-error edge case, full cloud login including `endpoint` → `transactor_base` override, OTP send/validate flow, and result-hardening regressions. Full suite: 256 passed.

Let me know if you'd like anything restructured.

**Comment by teslakoile** [comment]
Thanks for the rework — all four prior blockers look properly addressed and the new test file is well-targeted. One ask before merge:

**Add command-level tests for `auth login-otp`** in `tests/test_auth_cmd.py`. The new command has non-trivial branches (`--code` provided vs. send-then-prompt, `AuthError` → `Exit(2)`, `Exception` → `Exit(1)`) that aren't covered. Mirror the existing `auth_login` CliRunner tests with `send_otp`/`login_otp` mocked.

Nits (fine in a follow-up):

- `_is_cloud` matches `*.huly.app` (with the leading dot), so a bare `https://huly.app` wouldn't classify as cloud. Probably intentional — a one-line comment would make it explicit.
- `login()` docstring still lists steps 2–5 inline, but those moved into `_complete_login`. Worth updating to point at the helper.
- `_complete_login` applies the `endpoint` override unconditionally. If a self-hosted instance ever populates `selectWorkspace.endpoint`, the `/_transactor` reverse-proxy path gets silently replaced. Consider gating the override on `_is_cloud(config)` unless the upstream contract guarantees the field is cloud-only.
- `HulyConfig.otp_code` is a transient runtime value sitting on the long-lived config dataclass. Passing it as an argument to `login_otp` would be cleaner, though `repr=False` + non-persistence already covers the practical concerns.

**Comment by fabioklr** [comment]
Thanks for the review — pushed four commits addressing all four points:

- [`c5d3ba9`](https://github.com/teslakoile/huly-cli/pull/13/commits/c5d3ba9) refactor — `login_otp(config, code)` takes the code as an arg; dropped the transient `HulyConfig.otp_code` field.
- [`c57eac1`](https://github.com/teslakoile/huly-cli/pull/13/commits/c57eac1) fix — gated the `selectWorkspace.endpoint` override on `_is_cloud(config)` so a self-hosted server populating that field can't accidentally bypass the `/_transactor` reverse-proxy. Added a regression test in `test_auth_cloud.py`.
- [`6b7b26d`](https://github.com/teslakoile/huly-cli/pull/13/commits/6b7b26d) docs — `login()` docstring now points at `_complete_login` instead of duplicating steps 2–5; added a one-liner on `_is_cloud` explaining the leading-dot match excludes bare `huly.app` by design.
- [`6ae5366`](https://github.com/teslakoile/huly-cli/pull/13/commits/6ae5366) test — six `CliRunner` tests for `auth login-otp` mirroring the existing `auth_login` style: `--code` provided (skips `send_otp`), no `--code` non-TTY, `send_otp` `AuthError` → exit 2, `login_otp` `AuthError` → exit 2, `login_otp` generic `Exception` → exit 1, and missing email/workspace non-TTY parity.

Full suite green locally (264 passed). Ready for another look.
